File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/playlist/ProviderClient.java
package app.zxtune.playlist;

import android.content.ContentResolver;
import android.content.Context;
import android.database.ContentObserver;
import android.database.Cursor;
import android.net.Uri;

import androidx.annotation.Nullable;

import java.util.HashMap;

import javax.annotation.CheckForNull;

import app.zxtune.analytics.Analytics;

public final class ProviderClient {

  //should be name-compatible with Database
  public enum SortBy {
    title,
    author,
    duration
  }

  public enum SortOrder {
    asc,
    desc
  }

  public interface ChangesObserver {
    void onChange();
  }

  private final ContentResolver resolver;
  @Nullable
  private ContentObserver contentObserver;

  private ProviderClient(Context ctx) {
    this.resolver = ctx.getContentResolver();
  }

  public static ProviderClient create(Context ctx) {
    return new ProviderClient(ctx);
  }

  public static Uri createUri(long id) {
    return PlaylistQuery.uriFor(id);
  }

  @Nullable
  public static Long findId(Uri uri) {
    return PlaylistQuery.isPlaylistUri(uri)
        ? PlaylistQuery.idOf(uri)
        : null;
  }

  public void addItem(Item item) {
    resolver.insert(PlaylistQuery.ALL, item.toContentValues());
  }

  public void notifyChanges() {
    resolver.notifyChange(PlaylistQuery.ALL, null);
  }

  public void registerObserver(final ChangesObserver observer) {
    if (contentObserver != null) {
      throw new IllegalStateException();
    }
    contentObserver = new ContentObserver(null) {
      @Override
      public boolean deliverSelfNotifications() {
        return false;
      }

      @Override
      public void onChange(boolean selfChange) {
        observer.onChange();
      }
    };
    resolver.registerContentObserver(PlaylistQuery.ALL, true, contentObserver);
  }

  public void unregisterObserver() {
    if (contentObserver != null) {
      resolver.unregisterContentObserver(contentObserver);
      contentObserver = null;
    }
  }

  @Nullable
  public Cursor query(@Nullable long[] ids) {
    return resolver.query(PlaylistQuery.ALL, null,
        PlaylistQuery.selectionFor(ids), null, null);
  }

  public void delete(long[] ids) {
    deleteItems(PlaylistQuery.selectionFor(ids));
    Analytics.sendPlaylistEvent(Analytics.PlaylistAction.DELETE, ids.length);
  }

  public void deleteAll() {
    deleteItems(null);
    Analytics.sendPlaylistEvent(Analytics.PlaylistAction.DELETE, 0);
  }

  private void deleteItems(@Nullable String selection) {
    resolver.delete(PlaylistQuery.ALL, selection, null);
    notifyChanges();
  }

  public void move(long id, int delta) {
    Provider.move(resolver, id, delta);
    notifyChanges();
    Analytics.sendPlaylistEvent(Analytics.PlaylistAction.MOVE, 1);
  }

  public void sort(SortBy by, SortOrder order) {
    Provider.sort(resolver, by.name(), order.name());
    notifyChanges();
    Analytics.sendPlaylistEvent(Analytics.PlaylistAction.SORT,
        100 * by.ordinal() + order.ordinal());
  }

  @Nullable
  public Statistics statistics(@Nullable long[] ids) {
    final Cursor cursor = resolver.query(PlaylistQuery.STATISTICS, null,
        PlaylistQuery.selectionFor(ids), null, null);
    if (cursor == null) {
      return null;
    }
    try {
      if (cursor.moveToFirst()) {
        return new Statistics(cursor);
      }
    } finally {
      cursor.close();
    }
    return null;
  }

  // id => path
  public HashMap<String, String> getSavedPlaylists(@CheckForNull String id) {
    final HashMap<String, String> result = new HashMap<>();
    final Cursor cursor = resolver.query(PlaylistQuery.SAVED, null, id, null, null);
    if (cursor != null) {
      try {
        while (cursor.moveToNext()) {
          result.put(cursor.getString(0), cursor.getString(1));
        }
      } finally {
        cursor.close();
      }
    }
    return result;
  }

  public void savePlaylist(String id, @CheckForNull long[] ids) throws Exception {
      Provider.save(resolver, id, ids);
      Analytics.sendPlaylistEvent(Analytics.PlaylistAction.SAVE, ids != null ? ids.length : 0);
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/playlist/ProviderClient.kt
package app.zxtune.playlist

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import androidx.annotation.VisibleForTesting
import app.zxtune.analytics.Analytics
import app.zxtune.playlist.IO.asStatistics
import app.zxtune.playlist.IO.toContentValues
import app.zxtune.playlist.IO.toPlaylist
import app.zxtune.playlist.IO.toTrack
import app.zxtune.ui.utils.observeChanges
import app.zxtune.ui.utils.query
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.transform
import kotlinx.coroutines.withContext

class PlaylistContent(size: Int) : ArrayList<Track>(size)

class ProviderClient @VisibleForTesting constructor(
    private val resolver: ContentResolver,
    val id: Playlist.Id,
    private val dispatcher: CoroutineDispatcher,
) {
    private val playlistUri = Query.localPlaylistUri(id)

    fun add(track: Track.Metadata) = resolver.insert(playlistUri, track.toContentValues())

    fun notifyChanges() = resolver.notifyChange(playlistUri, null)

    fun observeContent() = observeChanges().transform {
        queryContent()?.let {
            emit(it)
        }
    }

    @OptIn(FlowPreview::class)
    fun observeChanges() = resolver.observeChanges(playlistUri).debounce(1000)

    suspend fun queryContent() = resolver.query(playlistUri) { cursor ->
        PlaylistContent(cursor.count).apply {
            while (cursor.moveToNext()) {
                add(cursor.toTrack())
            }
        }
    }

    suspend fun delete(tracks: Track.IdSet) {
        deleteItems(tracks)
        Analytics.sendPlaylistEvent(Analytics.PlaylistAction.DELETE, tracks.size)
    }

    suspend fun clear() {
        deleteItems(null)
        Analytics.sendPlaylistEvent(Analytics.PlaylistAction.DELETE, 0)
    }

    private suspend fun deleteItems(tracks: Track.IdSet?) = withContext(dispatcher) {
        Provider.delete(resolver, Playlist.OperationScope(id, tracks))
        notifyChanges()
    }

    suspend fun reorder(track: Track.Id, delta: Int) = withContext(dispatcher) {
        Provider.reorder(resolver, Track.FullIdentifier(id, track), delta)
        notifyChanges()
        Analytics.sendEvent("ui/playlist/reorder", "delta" to delta)
    }

    suspend fun sort(spec: Playlist.Sorting) = withContext(dispatcher) {
        Provider.sort(resolver, id, spec)
        notifyChanges()
        Analytics.sendEvent(
            "ui/playlist/sort",
            "by" to spec.by.name.lowercase(),
            "order" to spec.order.name.lowercase()
        )
    }

    suspend fun statistics(tracks: Track.IdSet?) = withContext(dispatcher) {
        Analytics.sendPlaylistEvent(Analytics.PlaylistAction.STATISTICS, tracks?.size ?: 0)
        Provider.statistics(resolver, Playlist.OperationScope(id, tracks))?.asStatistics()
    }

    @Throws(Exception::class)
    suspend fun save(name: String, tracks: Track.IdSet?) = withContext(dispatcher) {
        Provider.save(resolver, name, Playlist.OperationScope(id, tracks))
        Analytics.sendPlaylistEvent(Analytics.PlaylistAction.SAVE, tracks?.size ?: 0)
    }

    companion object {
        @JvmStatic
        fun create(ctx: Context) = ProviderClient(ctx.contentResolver, Playlist.DEFAULT_ID, Dispatchers.IO)

        fun createUri(id: Track.FullIdentifier): Uri = Query.itemUri(id)

        fun findId(uri: Uri) = Query.findTrackId(uri)
    }
}

class AggregatingProviderClient(private val resolver: ContentResolver, private val dispatcher: CoroutineDispatcher) {
    @OptIn(FlowPreview::class)
    fun observePlaylists() = resolver.observeChanges(Query.savedPlaylistsUri()).debounce(1000).transform {
        queryPlaylists()?.let {
            emit(it)
        }
    }

    @VisibleForTesting
    suspend fun queryPlaylists() = resolver.query(Query.savedPlaylistsUri()) { cursor ->
        ArrayList<Playlist>(cursor.count).apply {
            while (cursor.moveToNext()) {
                add(cursor.toPlaylist())
            }
        }
    }

    fun getPlaylist(playlist: Playlist.Id) = ProviderClient(resolver, playlist, dispatcher)

    // id => path
    suspend fun getSavedPlaylists(id: String? = null) = withContext(dispatcher) {
        resolver.query(Query.savedPlaylistsUri(), selection = id) { cursor ->
            HashMap<String, String>().apply {
                while (cursor.moveToNext()) {
                    put(cursor.getString(0), cursor.getString(1))
                }
            }
        }
    }

    companion object {
        fun create(ctx: Context) = AggregatingProviderClient(ctx.contentResolver, Dispatchers.IO)
    }
}
